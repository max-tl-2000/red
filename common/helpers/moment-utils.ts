/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import moment, { Moment } from 'moment-timezone'; // eslint-disable-line red/no-moment
import * as baseMoment from 'moment'; // eslint-disable-line red/no-moment
import { extendMoment } from 'moment-range';
import nullish from './nullish';
import { isString } from './type-of';
import { isNode } from './globals';

export type MomentType = Moment | string | Date;

interface IIsSameDayOptions {
  timezone?: string;
}

interface ILogWarningsArgs {
  msg?: any[];
}

interface IToMomentArgs {
  parseFormat?: string;
  timezone?: string;
  strict?: boolean;
}

interface IFormatMomentArgs {
  format?: string;
  timezone?: string;
  includeZone?: boolean;
  includeZoneStrict?: boolean;
}

interface INowArgs {
  timezone?: string;
}

interface IParseASInTimezoneArgs {
  format?: string;
  timezone?: string;
}

const runningInNode = isNode();

const momentFn = moment;
const { duration, max: momentMax, now: momentNow } = momentFn;

export const DATE_ISO_FORMAT = 'YYYY-MM-DD';

// this format is not correct the UTC formats with `Z` don't include +z
export const DATE_TIME_ISO_FORMAT = 'YYYY-MM-DDTHH:mm:ss+z';

export const findLocalTimezone = (): string => momentFn.tz.guess();
export const isValidTimezone = (tz: string): boolean => !!momentFn.tz.zone(tz);

let momentUtilsFixedNow;

const logWarnings = ({ msg = [] }: ILogWarningsArgs = {}) => {
  const isProduction = process.env.NODE_ENV === 'production';

  if (isProduction || !msg || msg.length === 0) {
    return;
  }

  const isDevelopment = process.env.NODE_ENV === 'development';

  console.warn(...msg);

  if (isDevelopment && runningInNode) {
    const err = new Error();
    const LINES_TO_SKIP = 2;
    // we remove the 2 first lines which are meaningless for us
    // and also the lines that contains stack info from teh asyncGenerator babel transform
    const stackWithMeaningFulInfo = (err.stack || '')
      .split(/\n/)
      .slice(LINES_TO_SKIP)
      .filter(line => !line.match(/asyncToGenerator\.js/))
      .join('\n');

    console.warn(stackWithMeaningFulInfo);
  }
};

const checkIfDateIsISO = (date: MomentType): void => {
  const isProduction = process.env.NODE_ENV === 'production';

  if (isProduction || !isString(date)) {
    return;
  }

  date = date as string; // cast needed to make typescript happy

  const ISO_8601_REGEX_FULL = /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(\.\d+)?(([+-]\d\d:\d\d)|Z)?$/i;
  const ISO_8601_REGEX_PERMISSIVE = /^\d{4}(-\d\d(-\d\d(T\d\d:\d\d(:\d\d)?(\.\d+)?(([+-]\d\d:\d\d)|Z)?)?)?)?$/i;

  if (date.match(ISO_8601_REGEX_FULL)) {
    return;
  }

  if (date.match(ISO_8601_REGEX_PERMISSIVE)) {
    // if it matches the permissive then we just inform it failed the full ISO validation
    logWarnings({ msg: ['>> `date` parameter is not in full ISO format', date] });
    return;
  }

  logWarnings({ msg: ['>> date parameter is not in any ISO format', date] });
};

export const toMoment = (date: MomentType, { parseFormat, timezone, strict = true }: IToMomentArgs = { strict: true }): Moment => {
  if (nullish(date) || date === '') {
    logWarnings({
      msg: ['>> date parameter is nullish or empty', date, '\n', '>> this will produce an invalid moment that will be formatted as empty string'],
    });

    const m = momentFn(null!);
    m.format = () => '';
    return m; // just return an invalid moment object
  }

  if (parseFormat !== undefined && typeof parseFormat !== 'string') {
    throw new TypeError('`parseFormat` must be a string');
  }

  checkIfDateIsISO(date);

  if (timezone !== undefined && !isString(timezone)) {
    throw new TypeError('timezone must be a string');
  }

  let dateAsMoment = parseFormat ? momentFn(date, parseFormat, strict) : momentFn(date);

  if (timezone) {
    dateAsMoment = dateAsMoment.tz(timezone);
  }

  return dateAsMoment;
};

export const formatMoment = (
  momentObj: Moment,
  { format, timezone, includeZone = true, includeZoneStrict = false }: IFormatMomentArgs = { includeZone: true, includeZoneStrict: false },
): string => {
  if (!timezone) {
    logWarnings({ msg: ['[formatMoment] called without timezone', momentObj] });
  }

  const m = momentObj ? toMoment(momentObj, { timezone }) : undefined;

  if (!m || !m.isValid()) {
    logWarnings({ msg: ['[formatMoment] attempt to format an invalid moment', momentObj] });
    return '';
  }

  const useAbbr = timezone && includeZone && timezone !== findLocalTimezone();

  const zone = useAbbr || includeZoneStrict ? ` ${m.zoneAbbr()}` : '';
  return `${m.format(format)}${zone}`;
};

export const now = ({ timezone = findLocalTimezone() }: INowArgs = { timezone: findLocalTimezone() }) => {
  if (momentUtilsFixedNow) {
    return momentUtilsFixedNow;
  }

  if (timezone !== undefined && typeof timezone !== 'string') {
    throw new TypeError('timezone must be a string');
  }

  let m = momentFn();
  if (timezone) {
    m = m.tz(timezone);
  }
  return m;
};

export const useFixedDateTimeForMomentUtilsNow = dateTime => {
  momentUtilsFixedNow = dateTime;

  return () => {
    momentUtilsFixedNow = undefined;
  };
};

export const parseAsInTimezone = (str: string, { format, timezone }: IParseASInTimezoneArgs = {}): Moment => {
  const args = [str];

  if (format) {
    args.push(format);
  }

  if (timezone) {
    args.push(timezone);
  } else {
    logWarnings({ msg: ['[parseAsInTimezone] error: missing `timezone`. In future versions this will throw an error'] });
  }

  // @ts-ignore: spread will work just fine in this case
  return momentFn.tz(...args);
};

export const isMoment = (m: any): boolean => momentFn.isMoment(m);

export const isSameDay = (date1, date2, { timezone }: IIsSameDayOptions = {}): boolean => {
  if (!timezone) {
    logWarnings({ msg: ['[isSameDay] called without timezone'] });
  }

  date1 = toMoment(date1, { timezone });
  date2 = toMoment(date2, { timezone });

  return !date1.isValid() || !date2.isValid() ? false : date1.isSame(date2, 'day');
};

export const isValidMoment = (m: any): boolean => m && isMoment(m) && m.isValid();

export const setDefaultTZ = (tz: string) => momentFn.tz.setDefault(tz);

export const getDurationBetweenMoments = (startDate: MomentType, endDate: MomentType): string => {
  const millisecondDifference = toMoment(endDate).diff(startDate);
  return moment.utc(millisecondDifference).format('mm:ss');
};

export const rangeMoment = extendMoment(baseMoment);

export const getTimezoneNames = (): Array<string> => momentFn.tz.names();

export const stringRepresentsADateWithoutTime = date => {
  if (typeof date !== 'string') {
    return false;
  }
  const POSIBLE_DATE = /^\d{4}-\d{1,2}-\d{1,2}$/g;
  return !!date.match(POSIBLE_DATE);
};

export const isWithinLastNDays = ({ date, days, timezone }) => toMoment(date, { timezone }).isSameOrAfter(now().add(days, 'days'), 'day');

export { duration, momentMax, momentNow };

export const today = ({ timezone = findLocalTimezone() }: INowArgs = { timezone: findLocalTimezone() }) =>
  now({ timezone }).set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
