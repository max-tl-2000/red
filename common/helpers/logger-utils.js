/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import produce, { setAutoFreeze } from 'immer';
import stringify from 'json-stringify-safe';
import set from 'lodash/set';
import has from 'lodash/has';
import get from 'lodash/get';
import unset from 'lodash/unset';
import pick from 'lodash/pick';
import envVal from './env-val';
import { isMoment } from './moment-utils';
import { isObject } from './type-of';
import { isUrl, getUrlParamsValueMatcher } from '../regex';

const isLogTransformEnabled = envVal('ENABLE_LOG_TRANSFORM', 'true') === 'true';

const excludedFieldsEntries = ['sendSsnEnabled'];

export const privateLogEntries = ['password', 'socSecNumber', 'api-token', 'authorization_token'];

const useLogConsole = envVal('ENABLE_LOG_CONSOLE', 'false') === 'true';
const logConsole = {
  log: (...args) => useLogConsole && console.log(...args),
  error: (...args) => useLogConsole && console.error(...args),
};

// the values specify the field to remain from the object.
// e.g. "quote": { "id": "123", "quoteContent": "555" } would be reduced to "quote": { "id": "123"}
// if no value is defined, the field will be omitted completely
const LOG_PATHS_TO_OMIT = [
  'messageEntity.message',
  'emailInfo.ctx',
  'deliveryResponses',
  'responses',
  'eventResponse',
  'inventories',
  'payload.msgCtx',
  'commData.message',
];

const LOG_FIELDS_TO_COLLAPSE = ['cache', 'quote:id'];

export const logCtx = ctx => {
  const { tenantId, reqId, msgId, routingKey, retryCount, documentVersion, originalRequestIds, authUser, trx } = ctx || {};

  const { userId, fullName } = authUser || {};
  const { trxId } = trx || {};

  const outValues = {
    tenantId,
    reqId,
    msgId,
    routingKey,
    retryCount,
    documentVersion,
    originalRequestIds,
  };

  const retVal = Object.keys(outValues).reduce((acc, key) => {
    if (outValues[key] !== undefined) {
      acc[key] = outValues[key];
    }
    return acc;
  }, {});

  if (userId || fullName) {
    const aUser = {};
    if (userId) {
      aUser.userId = userId;
    }
    if (fullName) {
      aUser.fullName = fullName;
    }
    retVal.authUser = aUser;
  }

  if (trxId) {
    retVal.trx = { trxId };
  }

  return retVal;
};

export const OBSCURE_VALUE = 'REDACTED';

export const obscureUrl = (url = '', obscureQueryParams = privateLogEntries, replaceValue = `$1${OBSCURE_VALUE}$2`) =>
  url.replace(getUrlParamsValueMatcher(obscureQueryParams), replaceValue);

const isExcludedField = key => excludedFieldsEntries.includes(key);
const replaceValue = (key, value, properties) => {
  if (!isExcludedField(key) && properties.some(property => property && (property === key || key.toLowerCase().includes(property)))) {
    return OBSCURE_VALUE;
  }

  if (key.toLowerCase().includes('url') && isUrl(value)) {
    return obscureUrl(value);
  }

  return value;
};

const redactValuesInObjectIfNeeded = (obj, properties) => {
  if (!isObject(obj)) return obj;
  const keys = Object.keys(obj || {});

  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    const value = obj[key];
    if (!Array.isArray(value) && !isObject(value)) {
      obj[key] = replaceValue(key, value, properties);
    }

    if (isObject(value)) {
      obj[key] = redactValuesInObjectIfNeeded(value, properties);
    }

    if (Array.isArray(value)) {
      obj[key] = value.map(val => redactValuesInObjectIfNeeded(val, properties));
    }
  }

  return obj;
};

export const obscureObject = (obj, properties = privateLogEntries) => {
  if (!(obj && properties.length)) return obj;
  // we should consider to remove this as this makes the returned draft
  // not to be frozen, but our code currently modifies the returned object
  // from obscureObject in some places
  setAutoFreeze(false);
  return produce(obj, draft => redactValuesInObjectIfNeeded(draft, properties));
};

export const getSensitiveObject = obj => ({ sensitiveData: { ...obj } });

export const removeTokenFromObject = data => {
  const { token, ...rest } = data || {};
  return rest;
};

export const handleErrorObject = log => {
  const { err, error, ...rest } = log || {};
  return { ...rest, error: err || error };
};

// ensures that no field that is a valid Moment object ever gets serialized as an empty string
// This prevents ES from complaining when an empty string is passed into a mapped date field
const normalizeMomentField = (obj, key) => {
  const objValue = obj[key];
  if (isMoment(objValue)) {
    obj[key] = objValue.isValid() ? objValue.format('DD MMM, YYYY') : null;
    return true;
  }
  return false;
};

const stringifyIfArray = (obj, key) => {
  const objValue = obj[key];
  if (Array.isArray(objValue)) {
    obj[key] = stringify(objValue);
    return true;
  }
  return false;
};

// Removes any excluded paths  from the provided object, returning them in a map so they can be restored later
function removeExcludedPaths(obj) {
  const excludedObjs = [];
  // stash excludedPath objects
  LOG_PATHS_TO_OMIT.forEach(path => {
    if (has(obj, path)) {
      const val = get(obj, path);
      excludedObjs.unshift([path, val]);
      unset(obj, path);
    }
  });
  // logConsole.log(stringify({ excludedObjs }, null, 2));
  return excludedObjs;
}

const restoreModifiedPaths = (obj, modifiedPaths) => {
  // stash excludedPath objects
  modifiedPaths.forEach(([path, val]) => {
    if (val !== undefined) {
      set(obj, path, val);
    } else {
      unset(obj, path);
    }
  });
};

export const collapsedFieldReplacer = (key, value) => {
  // logConsole.log('collapsedFieldReplacer', { key, fullPath }); // stringifyvalue });
  const collapsedValue = LOG_FIELDS_TO_COLLAPSE.find(f => f.startsWith(key));

  if (collapsedValue && !!key) {
    logConsole.log('==== collapsedFieldReplacer found key', key); // , stringify(collapsedValue));
    const [_, replacementKey] = collapsedValue.split(':');
    logConsole.log('==== collapsedFieldReplacer found key', key); // , stringify(collapsedValue), replacementKey);
    return replacementKey === undefined ? undefined : pick(value, replacementKey);
  }
  return value;
};

// parentPath is array of keys qleading up to the root object
export const normalizeFields = (obj, parentPath = []) => {
  // console.log('normalizeFields', parentPath);
  const depth = parentPath.length;
  const savedFields = [];
  const fieldsToOmit = [];
  if (obj === null) {
    logConsole.log('normalizeFields got passed null obj - nothing to do');
    return [];
  }
  Object.keys(obj).forEach(key => {
    const objValue = obj[key];
    const fullPath = [...parentPath, key].join('.');
    // console.log(fullPath);
    let collapsedField;
    logConsole.log('handling key', fullPath, ' at depth', depth);
    logConsole.log(LOG_FIELDS_TO_COLLAPSE.find(f => f.startsWith(key)));
    if (normalizeMomentField(obj, key) || stringifyIfArray(obj, key)) {
      // console.log('pushing specifal field')
      savedFields.push([fullPath, objValue]);
    } else if (LOG_PATHS_TO_OMIT.includes(fullPath)) {
      // logConsole.log('PUSHING EXCLUDED FIELD');
      fieldsToOmit.push([key, objValue, fullPath]);
    } else if ((collapsedField = LOG_FIELDS_TO_COLLAPSE.find(f => f.startsWith(key)))) {  // eslint-disable-line
      // console.log('=====found omitable field', key, fullPath);
      const [_, replacementField] = collapsedField.split(':');
      // console.log('replacementField is',replacementField);
      if (replacementField) {
        // console.log('setting from replacementField');
        savedFields.push([fullPath, objValue]);
        obj[key] = objValue[replacementField];
      } else {
        // console.log('pushed key to fieldsToOmit');
        fieldsToOmit.push([key, obj[key], fullPath]);
      }
    } else if (typeof objValue === 'object') {
      // console.log('handling object at depth', depth);
      if (!depth) {
        // console.log('calling normalizeFields for obj');
        // logConsole.log('adding saved field for top-level object', fullPath, key);
        const newSavedField = normalizeFields(objValue, [...parentPath, key]);
        if (newSavedField.length) {
          savedFields.push(...newSavedField);
        }
      } else {
        // console.log('stringifying deep obj', key, fullPath);
        obj[key] = stringify(objValue, collapsedFieldReplacer);
        // console.log('back from stringifying deep obj', key, fullPath);
        // console.log('obj is now', stringify(obj));
        savedFields.push([fullPath, objValue]);
      }
    }
  });
  // console.log('clearing omitted fields');
  fieldsToOmit.forEach(([key, restoreValue, fullPath]) => {
    logConsole.log('OMITTING', key, fullPath);
    unset(obj, key);
    // logConsole.log(stringify(obj));
    savedFields.push([fullPath, restoreValue]);
  });

  // console.log('normalize returning savedFields')
  return savedFields;
};

// mutates the passed in "logs" object by:
// 1 - removing excluded paths
// 2 - replacing quote objects with quote ID
// 3 - stringifying any objects at level 2 or below
// 4 - stringifying arrays
// 5 - normalizing moment objects
// Because we mutate the passed in object, we have to restore it
// after the logs are emitted.  This is done by returning a restorer function to be called by the emit override
// after it has called the original emit function.
export const formatLogs = logs => {
  const pathsToRestore = [];
  const restorer = () => restoreModifiedPaths(logs, pathsToRestore);
  try {
    if (isLogTransformEnabled) {
      pathsToRestore.push(...removeExcludedPaths(logs), ...normalizeFields(logs));
    }
  } catch (e) {
    logConsole.log('ERROR in log transform!');
    logConsole.error('Error while transforming log entry', e);
    logs = { ...logs, orgMsg: logs.msg, orgLevel: logs.level, msg: 'UNFORMATTABLE LOG', loggingError: e.stack, level: 50 /* error */ };
  }
  return restorer;
};
