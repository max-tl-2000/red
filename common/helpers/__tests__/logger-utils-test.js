/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { obscureObject, normalizeFields } from '../logger-utils';
import { toMoment } from '../moment-utils';
import { LA_TIMEZONE } from '../../date-constants';

const propertiesToObscure = ['socSecNumber', 'token', 'Token', 'key', 'Key', 'authorization', 'Authorization', 'ssn', 'password'];

describe('logger-util-test', () => {
  describe('obscureObject', () => {
    it('should obscure the fields on an object', () => {
      const obj = {
        socSecNumber: 'socSecNumber',
        token: 'token',
        Token: 'Token',
        key: 'key',
        Key: 'Key',
        authorization: 'authorization',
        Authorization: 'Authorization',
        ssn: 'ssn',
        password: 'some password',
      };

      const result = obscureObject(obj, propertiesToObscure);

      expect(result).toMatchSnapshot();
    });

    it('should obscure the fields on an object except the ones that do not match the properties to obscure', () => {
      const obj = {
        socSecNumber: 'socSecNumber',
        token: 'token',
        Token: 'Token',
        key: 'key',
        Key: 'Key',
        authorization: 'authorization',
        Authorization: 'Authorization',
        ssn: 'ssn',
        password: 'some password',
        anotherField: 'some field',
        another: {
          field: {
            foo: 'bar',
          },
        },
      };

      const result = obscureObject(obj, propertiesToObscure);

      expect(result).toMatchSnapshot();
    });

    it('should obscure properties on nested objects', () => {
      const obj = {
        obscureMe: {
          ifYouCan: {
            whenNested: {
              socSecNumber: 'socSecNumber',
              token: 'token',
              Token: 'Token',
              key: 'key',
              Key: 'Key',
              authorization: 'authorization',
              Authorization: 'Authorization',
              ssn: 'ssn',
              password: 'some password',
              exceptMe: 'I should not be redacted',
            },
          },
        },
        anotherField: 'some field',
        another: {
          field: {
            foo: 'bar',
          },
        },
      };

      const result = obscureObject(obj, propertiesToObscure);

      expect(result).toMatchSnapshot();
    });

    it('should obscure properties on arrays as well', () => {
      const obj = {
        obscureMe: {
          ifYouCan: {
            whenNestedAsArray: [
              { socSecNumber: 'socSecNumber' },
              { token: 'token' },
              { Token: 'Token' },
              { key: 'key' },
              { Key: 'Key' },
              { authorization: 'authorization' },
              { Authorization: 'Authorization' },
              { ssn: 'ssn' },
              { password: 'some password', exceptMe: 'I should not be redacted' },
              { exceptMe: 'I should not be redacted' },
              'I should not be changed',
              'Or me since we are strings',
              { IShouldNotBeRedacted: { foo: 'bar' } },
            ],
          },
        },
        anotherField: 'some field',
        another: {
          field: {
            foo: 'bar',
          },
        },
        anArray: [
          { socSecNumber: 'socSecNumber' },
          { token: 'token' },
          { Token: 'Token' },
          { key: 'key' },
          { Key: 'Key' },
          { authorization: 'authorization' },
          { Authorization: 'Authorization' },
          { ssn: 'ssn' },
          { password: 'some password', exceptMe: 'I should not be redacted' },
          { exceptMe: 'I should not be redacted' },
          'I should not be changed',
          'Or me since we are strings',
          { IShouldNotBeRedacted: { foo: 'bar' } },
        ],
      };

      const result = obscureObject(obj, propertiesToObscure);

      expect(result).toMatchSnapshot();
    });

    describe('when null or undefined is provided', () => {
      it('should return null', () => {
        const obj = null;

        let result = obscureObject(obj, propertiesToObscure);
        expect(result).toEqual(null);

        result = obscureObject(undefined, propertiesToObscure);
        expect(result).toEqual(undefined);
      });
    });
  });

  describe('normalizeFields', () => {
    describe('when passing a nested object with 2 levels', () => {
      it('should return stringify the second level', () => {
        const obj = {
          levelId: 'level1',
          levelDifficulty: 'easy',
          nestedLevelTwo: {
            levelId: 'level2',
            levelDifficulty: 'medium',
          },
        };

        normalizeFields(obj);
        expect(obj).toMatchSnapshot();
      });
    });
    describe('when passing a nested object with 3 levels or more', () => {
      it('should return stringify the second level and beyond', () => {
        const obj = {
          levelId: 'level1',
          levelDifficulty: 'easy',
          nestedLevelTwo: {
            levelId: 'level2',
            levelDifficulty: 'medium',
            nestedLevelThree: {
              levelId: 'level3',
              levelDifficulty: 'hard',
              nestedLevelFour: {
                levelId: 'level4',
                levelDifficulty: 'extremly hard',
                nestedLevelFive: {
                  levelId: 'level5',
                  levelDifficulty: 'Stormy Ascent',
                },
              },
            },
          },
        };

        normalizeFields(obj);
        expect(obj).toMatchSnapshot();
      });
    });
    describe('when passing a nest object with multiple 2 levels', () => {
      it('should return stringify all second level', () => {
        const obj = {
          levelId: 'level1',
          levelDifficulty: 'easy',
          nestedLevelTwo: {
            levelId: 'level2',
            levelDifficulty: 'medium',
            nestedLevelThreeV1: {
              levelId: 'level3v1',
              levelDifficulty: 'hard',
            },
            nestedLevelTwoV2: {
              levelId: 'level2v2',
              levelDifficulty: 'medium',
            },
          },
        };

        normalizeFields(obj);
        expect(obj).toMatchSnapshot();
      });
    });
    describe('when passing an object with a valid moment object', () => {
      it('should return the moment property formated', () => {
        const momentValue = toMoment('2021-01-28T05:00:00Z', { timezone: LA_TIMEZONE });

        const obj = {
          levelId: 'level1',
          levelDifficulty: 'easy',
          momentValue,
        };

        normalizeFields(obj);
        expect(obj).toMatchSnapshot();
      });
    });
    describe('when passing an object with an invalid moment object', () => {
      it('should return the moment property as null', () => {
        const momentValue = toMoment('', { timezone: LA_TIMEZONE });

        const obj = {
          levelId: 'level1',
          levelDifficulty: 'easy',
          momentValue,
        };

        normalizeFields(obj);
        expect(obj).toMatchSnapshot();
      });
    });
  });

  describe('when passing an object with a nested quote object', () => {
    it('it should remove all quote object properties except the id as quoteId, and remove any cache objects', () => {
      const obj = {
        quote: { name: 'quoteAtTopLevel', id: 'quoteTopLevelId' },
        levelId: 'level1',
        levelDifficulty: 'easy',
        nestedLevelTwo: {
          levelId: 'level2',
          levelDifficulty: 'medium',
          cache: 'cacheAtLevel2',
        },
        cache: 'topLevelCache',
        person: { name: 'yoda', isYedi: true },
        application: {
          quote: { id: 2, name: 'name', cache: 'cacheInQuote' },
          cache: 'cacheInApplication',
          deepQuote: { deepQuoteId: '1234', quote: { id: 3, name: 'deepQuoteName', cache: 'cacheInDeepQuote' } },
        },
      };

      normalizeFields(obj); // this mutates obj
      expect(obj).toMatchSnapshot();

      // TODO: test to verify restorer function can be used to restore obj to its original sttate
    });
  });
});
