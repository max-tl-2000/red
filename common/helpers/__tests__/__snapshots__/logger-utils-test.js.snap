// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger-util-test normalizeFields when passing a nest object with multiple 2 levels should return stringify all second level 1`] = `
Object {
  "levelDifficulty": "easy",
  "levelId": "level1",
  "nestedLevelTwo": Object {
    "levelDifficulty": "medium",
    "levelId": "level2",
    "nestedLevelThreeV1": "{\\"levelId\\":\\"level3v1\\",\\"levelDifficulty\\":\\"hard\\"}",
    "nestedLevelTwoV2": "{\\"levelId\\":\\"level2v2\\",\\"levelDifficulty\\":\\"medium\\"}",
  },
}
`;

exports[`logger-util-test normalizeFields when passing a nested object with 2 levels should return stringify the second level 1`] = `
Object {
  "levelDifficulty": "easy",
  "levelId": "level1",
  "nestedLevelTwo": Object {
    "levelDifficulty": "medium",
    "levelId": "level2",
  },
}
`;

exports[`logger-util-test normalizeFields when passing a nested object with 3 levels or more should return stringify the second level and beyond 1`] = `
Object {
  "levelDifficulty": "easy",
  "levelId": "level1",
  "nestedLevelTwo": Object {
    "levelDifficulty": "medium",
    "levelId": "level2",
    "nestedLevelThree": "{\\"levelId\\":\\"level3\\",\\"levelDifficulty\\":\\"hard\\",\\"nestedLevelFour\\":{\\"levelId\\":\\"level4\\",\\"levelDifficulty\\":\\"extremly hard\\",\\"nestedLevelFive\\":{\\"levelId\\":\\"level5\\",\\"levelDifficulty\\":\\"Stormy Ascent\\"}}}",
  },
}
`;

exports[`logger-util-test normalizeFields when passing an object with a valid moment object should return the moment property formated 1`] = `
Object {
  "levelDifficulty": "easy",
  "levelId": "level1",
  "momentValue": "27 Jan, 2021",
}
`;

exports[`logger-util-test normalizeFields when passing an object with an invalid moment object should return the moment property as null 1`] = `
Object {
  "levelDifficulty": "easy",
  "levelId": "level1",
  "momentValue": null,
}
`;

exports[`logger-util-test obscureObject should obscure properties on arrays as well 1`] = `
Object {
  "anArray": Array [
    Object {
      "socSecNumber": "REDACTED",
    },
    Object {
      "token": "REDACTED",
    },
    Object {
      "Token": "REDACTED",
    },
    Object {
      "key": "REDACTED",
    },
    Object {
      "Key": "REDACTED",
    },
    Object {
      "authorization": "REDACTED",
    },
    Object {
      "Authorization": "REDACTED",
    },
    Object {
      "ssn": "REDACTED",
    },
    Object {
      "exceptMe": "I should not be redacted",
      "password": "REDACTED",
    },
    Object {
      "exceptMe": "I should not be redacted",
    },
    "I should not be changed",
    "Or me since we are strings",
    Object {
      "IShouldNotBeRedacted": Object {
        "foo": "bar",
      },
    },
  ],
  "another": Object {
    "field": Object {
      "foo": "bar",
    },
  },
  "anotherField": "some field",
  "obscureMe": Object {
    "ifYouCan": Object {
      "whenNestedAsArray": Array [
        Object {
          "socSecNumber": "REDACTED",
        },
        Object {
          "token": "REDACTED",
        },
        Object {
          "Token": "REDACTED",
        },
        Object {
          "key": "REDACTED",
        },
        Object {
          "Key": "REDACTED",
        },
        Object {
          "authorization": "REDACTED",
        },
        Object {
          "Authorization": "REDACTED",
        },
        Object {
          "ssn": "REDACTED",
        },
        Object {
          "exceptMe": "I should not be redacted",
          "password": "REDACTED",
        },
        Object {
          "exceptMe": "I should not be redacted",
        },
        "I should not be changed",
        "Or me since we are strings",
        Object {
          "IShouldNotBeRedacted": Object {
            "foo": "bar",
          },
        },
      ],
    },
  },
}
`;

exports[`logger-util-test obscureObject should obscure properties on nested objects 1`] = `
Object {
  "another": Object {
    "field": Object {
      "foo": "bar",
    },
  },
  "anotherField": "some field",
  "obscureMe": Object {
    "ifYouCan": Object {
      "whenNested": Object {
        "Authorization": "REDACTED",
        "Key": "REDACTED",
        "Token": "REDACTED",
        "authorization": "REDACTED",
        "exceptMe": "I should not be redacted",
        "key": "REDACTED",
        "password": "REDACTED",
        "socSecNumber": "REDACTED",
        "ssn": "REDACTED",
        "token": "REDACTED",
      },
    },
  },
}
`;

exports[`logger-util-test obscureObject should obscure the fields on an object 1`] = `
Object {
  "Authorization": "REDACTED",
  "Key": "REDACTED",
  "Token": "REDACTED",
  "authorization": "REDACTED",
  "key": "REDACTED",
  "password": "REDACTED",
  "socSecNumber": "REDACTED",
  "ssn": "REDACTED",
  "token": "REDACTED",
}
`;

exports[`logger-util-test obscureObject should obscure the fields on an object except the ones that do not match the properties to obscure 1`] = `
Object {
  "Authorization": "REDACTED",
  "Key": "REDACTED",
  "Token": "REDACTED",
  "another": Object {
    "field": Object {
      "foo": "bar",
    },
  },
  "anotherField": "some field",
  "authorization": "REDACTED",
  "key": "REDACTED",
  "password": "REDACTED",
  "socSecNumber": "REDACTED",
  "ssn": "REDACTED",
  "token": "REDACTED",
}
`;

exports[`logger-util-test when passing an object with a nested quote object it should remove all quote object properties except the id as quoteId, and remove any cache objects 1`] = `
Object {
  "application": Object {
    "deepQuote": "{\\"deepQuoteId\\":\\"1234\\",\\"quote\\":{\\"id\\":3}}",
    "quote": 2,
  },
  "levelDifficulty": "easy",
  "levelId": "level1",
  "nestedLevelTwo": Object {
    "levelDifficulty": "medium",
    "levelId": "level2",
  },
  "person": Object {
    "isYedi": true,
    "name": "yoda",
  },
  "quote": "quoteTopLevelId",
}
`;
